### ISSUE TITLE
Sometimes imfile can't notice a file change on log

### ISSUE THREAD

**Issue by timp87** [rsyslog/rsyslog#1417]
Hello, I'm here again.
Please, do not relate this ticket to the previous mine. I'll describe exactly what I have and check everything at least 2 times.

We have a java application which writes and rotates log files by itself using Logback.
Logback is configured to rotate file it size is more than 100MB or if new day has come.

Rsyslog reads logfiles by imfile.
When we used rsyslog-8.19 we had no problems. Now we are trying to upgrade to the latest rsyslog-8.24 and have some problems.
Imfile can't notice a file change (rotation) on logfile sometimes.
This is how it looks like when log have just been rotated:
```
# lsof -p `pgrep rsyslog` | grep var/log
COMMAND    PID USER   FD      TYPE             DEVICE SIZE/OFF       NODE NAME
rsyslogd 19781 root   10w      REG              253,1   2102605         84 /var/log/spool/rsyslog/queue-sumologic-pas-imfile.00000001
rsyslogd 19781 root   11r      REG              253,1   2102605         84 /var/log/spool/rsyslog/queue-sumologic-pas-imfile.00000001
rsyslogd 19781 root   12r      REG              253,1   2102605         84 /var/log/spool/rsyslog/queue-sumologic-pas-imfile.00000001
rsyslogd 19781 root   18r      REG              253,1  79249872     819259 /var/log/mydomain/pas/pas.statistics-api.log
rsyslogd 19781 root   19w      REG              253,1  36774237        188 /var/log/messages
rsyslogd 19781 root   21w      REG              253,1  33887387         44 /var/log/kern.log
rsyslogd 19781 root   26r      REG              253,1 104866963     819287 /var/log/mydomain/pas/pas.main.log404357117420449.tmp (deleted)
rsyslogd 19781 root   27w      REG              253,1    144106        190 /var/log/secure
rsyslogd 19781 root   29w      REG              253,1   7638999        183 /var/log/syslog.log
```
The old pas.main.log has inode number 819287. New pas.main.log has 819286:
```
# stat /var/log/mydomain/pas/pas.main.log
  File: ‘/var/log/mydomain/pas/pas.main.log’
  Size: 29381957        Blocks: 57392      IO Block: 4096   regular file
Device: fd01h/64769d    Inode: 819286      Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/     pas)   Gid: ( 1000/     pas)
Access: 2017-02-14 13:56:34.680916867 +0000
Modify: 2017-02-14 13:59:14.976015154 +0000
Change: 2017-02-14 13:59:14.976015154 +0000
 Birth: -
```
Rsyslog can't notice it.
Rotation happened somewhere at 13:56:34 of 2017-02-14.

Configuration is:
```
# cat /etc/rsyslog.conf 

global(
    action.reportSuspension="on"
    net.enableDNS="off"
    net.ipprotocol="ipv4-only"
    senders.keepTrack="on"
    senders.reportGoneAway="on"
    senders.reportNew="on"
)

timezone(id="UTC" offset="+00:00")

$MaxOpenFiles 65535

module(load="impstats" format="cee" interval="60" resetCounters="on" severity="7" facility="6" log.syslog="on" ruleset="impstats")

$ModLoad immark

## provides UDP syslog reception
module(load="imudp")
input(type="imudp" address="0.0.0.0" port="514" rcvbufSize="1m" RateLimit.Interval="0" name="udp-server")

## provides TCP syslog reception
module(load="imtcp" MaxSessions="500" FlowControl="off" MaxListeners="32" )
input(type="imtcp" address="0.0.0.0" port="514" name="tcp-server")

## provides kernel logging support.
$ModLoad imklog
$KLogPermitNonKernelFacility on

## provides support for local system logging
module(load="imuxsock" SysSock.RateLimit.Interval="0" SysSock.UsePIDFromSystem="on")
input(type="imuxsock" Socket="/dev/log" RateLimit.Interval="0" UsePIDFromSystem="on" )

$WorkDirectory /var/log/spool/rsyslog

$MainMsgQueueFileName rsyslog-main-queue
$MainMsgQueueType Direct
$MainMsgQueueSaveOnShutdown on
$MainMsgQueueTimeoutShutdown 120000 # number is timeout in ms - 120sec
$MainMsgQueueTimeoutEnqueue 0 #We do not like throtteling, the message will then immediately be discarded.

# log every message
$RepeatedMsgReduction off

# Multiline support:
$EscapeControlCharactersOnReceive off

# Increased maximum message size:
$MaxMessageSize 640k

$LogRSyslogStatusMessages on

$ActionFileDefaultTemplate RSYSLOG_RCFormat
$ActionForwardDefaultTemplate RSYSLOG_RCFormat

$template RSYSLOG_RCFormat, "%TIMESTAMP:::date-rfc3339% %HOSTNAME% %syslogseverity-text:::uppercase% %syslogtag%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%\n"

# include additional config files
$IncludeConfig /etc/rsyslog.d/*.conf

###------------------- Permissions -------------------------------
##$umask 0000

$FileCreateMode 0600
$FileOwner root
$FileGroup root

$DirCreateMode 0755
$DirOwner root
$DirGroup root


##----------------- Logging Configuration ------------------------
# Log kernel messages
kern.*              action(name="kern" type="omfile" File="/var/log/kern.log")

*.info;mail,authpriv,cron.none      action(name="messages" type="omfile" File="/var/log/messages")

# The authpriv file has restricted access.
authpriv.*          action(name="secure" type="omfile" File="/var/log/secure")

# Log all the mail messages in one place.
mail.*              action(name="maillog" type="omfile" File="/var/log/maillog")

# Log cron stuff
cron.*              action(name="cron" type="omfile" File="/var/log/cron")

# Everybody gets emergency messages
*.emerg             action(name="emerg" type="omusrmsg" Users="*")

# Save news errors of level crit and higher in a special file.
uucp,news.crit      action(name="spooler" type="omfile" File="/var/log/spooler")

# Save boot messages also to boot.log
local7.*            action(name="boot" type="omfile" File="/var/log/boot.log")
```

Also what's included:
```
# cat /etc/rsyslog.d/001-rsyslog-remote.conf

$Modload mmsequence
$template RSYSLOG_RCTFormat, "%TIMESTAMP:::date-rfc3339% %HOSTNAME% %$!seq1% %syslogseverity-text:::uppercase% %syslogtag%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%\n"
$template RawFormat,"%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%"

*.*      call remote


ruleset(name="remote"
        queue.filename="queue-remote"
        queue.maxdiskspace="500m"
        queue.maxfilesize="10m"
        queue.type="FixedArray"
        queue.syncqueuefiles="off"
        queue.checkpointinterval="100000"
        queue.size="5000"
        queue.discardmark="4800"
        queue.dequeuebatchsize="1024"
        queue.timeoutenqueue="0"
)
{
    
 local1.* 
     action(type="mmsequence" 
     mode="key" 
     key="key1" 
     var="$!seq1" )
 
     action(type="omfwd" template="RSYSLOG_RCTFormat" protocol="tcp" target="ppc01-logstash01.mydomain.com" port="10001" 
     name="omfwd-remote-ppc01-logstash01.mydomain.com" 
     queue.FileName="omfwd-remote-ppc01-logstash01.mydomain.com" 
     action.resumeRetryCount="3" )

 local0.*  action(type="omfwd" template="RawFormat" protocol="tcp" target="ppc01-logstash01.mydomain.com" port="11001" 
     name="omfwd-remote-ppc01-logstash01.mydomain.com" 
     queue.FileName="omfwd-remote-ppc01-logstash01.mydomain.com" 
     action.resumeRetryCount="3" )

 local2.*  action(type="omfwd" template="RawFormat" protocol="tcp" target="ppc01-logstash01.mydomain.com" port="11002" 
     name="omfwd-remote-ppc01-logstash01.mydomain.com" 
     queue.FileName="omfwd-remote-ppc01-logstash01.mydomain.com" 
     action.resumeRetryCount="3" )

}
```
and
```
# cat /etc/rsyslog.d/002-sim.conf

*.info call sim

ruleset(name="sim"
        queue.filename="queue-sim"
        queue.maxdiskspace="500m"
        queue.maxfilesize="10m"
        queue.type="FixedArray"
        queue.workerthreads="1"
        queue.syncqueuefiles="off"
        queue.checkpointinterval="100000"
        queue.size="10000"
        queue.discardmark="9800"
        queue.dequeuebatchsize="1024"
        queue.timeoutenqueue="0"
)
{
*.info    action(type="omfwd" protocol="udp" target="10.13.36.116" port="514" 
      name="omfwd-sim-10.13.36.116" 
      action.resumeRetryCount="3" )
}
```
and
```
# cat /etc/rsyslog.d/009-sumo.conf

$template SumoRawFormat,"%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%"
$template SumoFormat,"%TIMESTAMP:::date-rfc3339%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%"

local0.*  call sumologic


ruleset(name="sumologic"
        queue.filename="queue-sumologic"
        queue.maxdiskspace="500m"
        queue.maxfilesize="10m"
        queue.type="FixedArray"
        queue.syncqueuefiles="off"
        queue.checkpointinterval="100000"
        queue.size="10000"
        queue.discardmark="9800"
        queue.dequeuebatchsize="1024"
        queue.timeoutenqueue="0"
)
{
    
local0.* action(type="omfwd" protocol="udp" target="ppc01-p03-lcs01.p03.mydomain.com" port="5050" 
      name="omfwd-sumo-raw-pas-udp-ppc01-p03-lcs01.p03.mydomain.com-5050" 
      queue.FileName="omfwd-sumo-raw-pas-udp-ppc01-p03-lcs01.p03.mydomain.com-5050" 
      template="SumoRawFormat" 
      action.resumeRetryCount="3" )

}
```
and
```
# cat /etc/rsyslog.d/01-facility_map_dynamic.conf

*.*   call rclogs

ruleset(name="rclogs"
        queue.filename="queue-rclogs"
        queue.maxdiskspace="500m"
        queue.maxfilesize="10m"
        queue.type="FixedArray"
        queue.syncqueuefiles="off"
        queue.checkpointinterval="100000"
        queue.size="10000"
        queue.discardmark="9800"
        queue.dequeuebatchsize="1024"
        queue.timeoutenqueue="0"
)
{

# RC pas server
$template RCLogFile-pas-default,"/var/log/mydomain/pas/pas-default.log"
$template RCLogFile-pas,"/var/log/mydomain/pas/pas-%app-name%.log"

if $syslogfacility-text == 'local0' and $app-name == '' then { 
     action(type="omfile" DynaFile="RCLogFile-pas-default" name="omfile-local0-default" )
} else if $syslogfacility-text == 'local0' then { 
     action(type="omfile" DynaFile="RCLogFile-pas" name="omfile-local0-dynaFile" )
}


}
```
and
```
# cat /etc/rsyslog.d/90-sumo-raw-pas.conf
module(load="imfile")

#pas.statistics-api.log
$InputFileBindRuleset sumologic-pas-imfile
$InputFileName /var/log/mydomain/pas/pas.statistics-api.log
$InputFileTag pas-statistics-api:
$InputFileStateFile pas-statistics-api
$InputFileReadMode 2
$InputFileSeverity notice
$InputFileFacility local0
$InputFilePersistStateInterval 100000
$InputRunFileMonitor

#pas.main.log
$InputFileBindRuleset sumologic-pas-imfile
$InputFileName /var/log/mydomain/pas/pas.main.log
$InputFileTag pas.main.log:
$InputFileStateFile pas-main-log
$InputFileReadMode 2
$InputFileSeverity notice
$InputFileFacility local1
$InputFilePersistStateInterval 100000
$InputRunFileMonitor 

#pas.statistics-sms.log
$InputFileBindRuleset sumologic-pas-imfile
$InputFileName /var/log/mydomain/pas/pas.statistics-sms.log
$InputFileTag pas-statistics-sms:
$InputFileStateFile pas-statistics-sms
$InputFileReadMode 2
$InputFileSeverity notice
$InputFileFacility local2
$InputFilePersistStateInterval 100000
$InputRunFileMonitor

ruleset(name="sumologic-pas-imfile"
        queue.filename="queue-sumologic-pas-imfile"
        queue.maxdiskspace="500m"
        queue.maxfilesize="10m"
        queue.type="FixedArray"
        queue.workerthreads="1"
        queue.syncqueuefiles="off"
        queue.checkpointinterval="100000"
        queue.size="5000"
        queue.discardmark="4800"
        queue.dequeuebatchsize="1024"
        queue.timeoutenqueue="0"
)
{
    
local0.*        action(type="omfwd" protocol="udp" target="ppc01-p03-lcs01.p03.mydomain.com" port="5050" 
      name="omfwd-sumologic-pas-imfile-udp-ppc01-p03-lcs01.p03.mydomain.com-5050" 
      queue.FileName="omfwd-sumologic-pas-imfile-udp-ppc01-p03-lcs01.p03.mydomain.com-5050" 
      template="SumoRawFormat" 
      action.resumeRetryCount="3" )

local1.*        action(type="omfwd" protocol="udp" target="ppc01-p03-lcs01.p03.mydomain.com" port="5051" 
      name="omfwd-sumologic-pas-imfile-udp-ppc01-p03-lcs01.p03.mydomain.com-5051" 
      queue.FileName="omfwd-sumologic-pas-imfile-udp-ppc01-p03-lcs01.p03.mydomain.com-5051" 
      template="SumoRawFormat" 
      action.resumeRetryCount="3" )

local2.*        action(type="omfwd" protocol="udp" target="ppc01-p03-lcs01.p03.mydomain.com" port="5052" 
      name="omfwd-sumologic-pas-imfile-udp-ppc01-p03-lcs01.p03.mydomain.com-5052" 
      queue.FileName="omfwd-sumologic-pas-imfile-udp-ppc01-p03-lcs01.p03.mydomain.com-5052" 
      template="SumoRawFormat" 
      action.resumeRetryCount="3" )

    
    *.*     call remote
}
```
and
```
# cat /etc/rsyslog.d/99-discard-local.conf

local0.*     stop


local0.*     stop

# discard OPS.syslog.sumologic.facility
local0.*     stop
```
and
```

# cat /etc/rsyslog.d/99-syslog.conf
module(load="omprog")

module(load="mmjsonparse") 
 template(name="stats" type="list") { 
 constant(value="{") 
 property(name="timereported" dateFormat="rfc3339" format="jsonf" outname="@timestamp") # the timestamp 
 constant(value=",") 
 property(name="hostname" format="jsonf" outname="host") # the host generating stats 
 constant(value=",\"source\":\"impstats\",") # hardcode "impstats" as a source 
 property(name="$!name" format="jsonf" outname="name") # queue name (Example: omfwd-remote-logstash.ops.mydomain.com) 
 constant(value=",") 
 property(name="$!processed" format="jsonf" outname="processed" mandatory="off") 
 constant(value=",") 
 property(name="$!failed" format="jsonf" outname="failed" mandatory="off") 
 constant(value=",") 
 property(name="$!suspended" format="jsonf" outname="suspended" mandatory="off") 
 constant(value=",") 
 property(name="$!resumed" format="jsonf" outname="resumed" mandatory="off") 
 constant(value=",") 
 property(name="$!enqueued" format="jsonf" outname="enqueued" mandatory="off") 
 constant(value=",") 
 property(name="$!full" format="jsonf" outname="full" mandatory="off") 
 constant(value=",") 
 property(name="$!discarded.full" format="jsonf" outname="discarded!full" mandatory="off") 
 constant(value=",") 
 property(name="$!discarded.nf" format="jsonf" outname="discarded!nf" mandatory="off") 
 constant(value="}") 
 constant(value="
") }

 ruleset(name="impstats"
    queue.type="Direct"
    queue.timeoutenqueue="0"
) {
    *.* action(name="syslog" type="omfile" File="/var/log/syslog.log")

         # PAS Component specific action 
 action( name="parse_impstats" type="mmjsonparse") # Parse JSON stats 
 *.* action(type="mmsequence" 
 mode="key" 
 key="key1" 
 var="$!seq1" ) 
 action(type="omfwd" protocol="tcp" target="ppc01-logstash01.mydomain.com" port="50000" 
 template="stats" 
 name="omfwd-remote-logstash.ops.mydomain.com" 
 queue.FileName="omfwd-remote-logstash.ops.mydomain.com" 
 action.resumeRetryCount="3" )
```
and finally
```
# cat /etc/rsyslog.d/listen.conf
input(type="imuxsock" Socket="/run/systemd/journal/syslog" RateLimit.Interval="0" UsePIDFromSystem="on" )
```
Versions are
```
# rpm -qa | grep rsyslog
rsyslog-mmjsonparse-8.24.0-2.el7.x86_64
rsyslog-8.24.0-2.el7.x86_64

# cat /etc/centos-release
CentOS Linux release 7.2.1511 (Core)
```
I've gathered debug output, which I'll provide a bit later, when I filter it.

**Comment by timp87** [comment]
Debug rsyslog output is here https://drive.google.com/file/d/0By9yUmYFrRAKTEhmWUF5cnlVLU0

**Comment by timp87** [comment]
Same behavior on rsyslog-8.25.0

**Comment by 7echno** [comment]
Hi, I did noticed similar behaviour during testing imfile and "copytrancate" option with logrotate.
Not sure if it's the same problem, as it seems that changing inode is not what copytrancate do, and don't know your logrotation setup.
But if logrotated file has some change immediately after logrotate occurs, I got some message lost or not reading this truncated file at all.
I did try also with "reopenOnTruncate" experimental option, but still seems to got problems. Hopefully I will get more time soon to test it so I will have some more accurate info on this.

Maybe you can try to set your logrotate with copytrancate, and try if you got lucky with "reopenOnTruncate". 
Just a suggestion.

**Comment by davidelang** [comment]
On Mon, 13 Mar 2017, 7echno wrote:

> Hi, I did noticed similar behaviour during testing imfile and "copytrancate" option with logrotate.
> Not sure if it's the same problem, as it seems that changing inode is not what copytrancate do, and don't know your logrotation setup.
> But if logrotated file has some change immediately after logrotate occurs, I got some message lost or not reading this truncated file at all.
> I did try also with "reopenOnTruncate" experimental option, but still seems to got problems. Hopefully I will get more time soon to test it so I will have some more accurate info on this.
>
> Maybe you can try to set your logrotate with copytrancate, and try if you got lucky with "reopenOnTruncate".
> Just a suggestion.

copytruncate does not change the inode, it makes a copy of the file and leaves 
the old file in place and trucates it. This is going to cause you to loose logs, 
both in rsyslog and in your rotated files. There are two race conditions.

1. between rsyslog reading the file and the truncation that makes rsyslog unable 
to get the latest entries written

2. after copying the file new logs can still be written to it before the 
truncation takes place, these logs will not be in the rotated file.

David Lang

**Comment by 7echno** [comment]
Hi David,

It's clear what "copytrance" do.
What I want to figure is if there is some issue with rsyslog in this simple scenario.

Version: rsyslog-8.24.0

rsyslog.conf

```
$WorkDirectory /var/lib/rsyslog
module(load="imfile")

input(
    type="imfile"
    reopenOnTruncate="on"
    PersistStateInterval="1"
    File="/var/log/test.input"
    Tag="test"
    Facility="local1"
)
local1.*    /var/log/test.log

```
/etc/logrotate.d/test
```
/var/log/test.input {
copytruncate
daily
rotate 10
missingok
notifempty
}
```
Sanity test, send 50 messages:
`for i in $(seq -w 1 50); do echo "test_message_$i" >> /var/log/test.input; done`

After few seconds:
```
# grep test_message /var/log/test.log | wc -l
50
```

And now goes the problematic part:
```
logrotate -f /etc/logrotate.d/test && for i in $(seq -w 1 50); do echo "test_message_after_$i" >> /var/log/test.input; done

2017-03-15T05:38:40.585678-04:00 qeos-176 test test_message_43
2017-03-15T05:38:40.585679-04:00 qeos-176 test test_message_44
2017-03-15T05:38:40.585756-04:00 qeos-176 test test_message_45
2017-03-15T05:38:40.585758-04:00 qeos-176 test test_message_46
2017-03-15T05:38:40.585880-04:00 qeos-176 test test_message_47
2017-03-15T05:38:40.585881-04:00 qeos-176 test test_message_48
2017-03-15T05:38:40.585963-04:00 qeos-176 test test_message_49
2017-03-15T05:38:40.585965-04:00 qeos-176 test test_message_50
2017-03-15T05:43:00.848824-04:00 qeos-176 test sage_after_37
2017-03-15T05:43:00.848832-04:00 qeos-176 test test_message_after_38
2017-03-15T05:43:00.848950-04:00 qeos-176 test test_message_after_39
2017-03-15T05:43:00.848952-04:00 qeos-176 test test_message_after_40
2017-03-15T05:43:00.849169-04:00 qeos-176 test test_message_after_41
2017-03-15T05:43:00.849171-04:00 qeos-176 test test_message_after_42
2017-03-15T05:43:00.855269-04:00 qeos-176 test test_message_after_43
2017-03-15T05:43:00.855272-04:00 qeos-176 test test_message_after_44
2017-03-15T05:43:00.855380-04:00 qeos-176 test test_message_after_45
2017-03-15T05:43:00.855382-04:00 qeos-176 test test_message_after_46
2017-03-15T05:43:00.855515-04:00 qeos-176 test test_message_after_47
2017-03-15T05:43:00.855517-04:00 qeos-176 test test_message_after_48
2017-03-15T05:43:00.855627-04:00 qeos-176 test test_message_after_49
2017-03-15T05:43:00.855633-04:00 qeos-176 test test_message_after_50
```

Everything works fine, if I put "sleep" of 6-7 seconds between force "logrotate" and "for loop".
Even with smaller sleep of few seconds, got the same results.

My guess that there is some issue with the offset, as if I got smaller number of messages in the "Sanity" part. Let's say 20. I will get something like missing first 14 messages + part of the 15th.

I was playing a little with it, and in my case this lost is happening only in the range of 1-69 sanity messages already in input file before logrotate. 
If I have more than 69 messages in "Sanity" part all messages after are logged without lost. Which was tricky as logs are usually bigger and seems that there is no problem then.   

Also I tried with polling and inotify mode., same issue.

**Comment by davidelang** [comment]
On Wed, 15 Mar 2017, 7echno wrote:

> Hi David,
>
> It's clear what "copytrance" do.
> What I want to figure is if there is some issue with rsyslog in this simple scenario.
>
> Version: rsyslog-8.24.0
>
> rsyslog.conf
>
> ```
> $WorkDirectory /var/lib/rsyslog
> module(load="imfile")
>
> input(
>    type="imfile"
>    reopenOnTruncate="on"
>    PersistStateInterval="1"
>    File="/var/log/test.input"
>    Tag="test"
>    Facility="local1"
> )
> local1.*    /var/log/test.log
>
> ```
> /etc/logrotate.d/test
> ```
> /var/log/test.input {
> copytruncate
> daily
> rotate 10
> missingok
> notifempty
> }
> ```
> Sanity test, send 50 messages:
> `for i in $(seq -w 1 50); do echo "test_message_$i" >> /var/log/test.input; done`
>
> After few seconds:
> ```
> # grep test_message /var/log/test.log | wc -l
> 50
> ```
>
> And now goes the problematic part:
> ```
> logrotate -f /etc/logrotate.d/test && for i in $(seq -w 1 50); do echo "test_message_after_$i" >> /var/log/test.input; done
>
> 2017-03-15T05:38:40.585678-04:00 qeos-176 test test_message_43
> 2017-03-15T05:38:40.585679-04:00 qeos-176 test test_message_44
> 2017-03-15T05:38:40.585756-04:00 qeos-176 test test_message_45
> 2017-03-15T05:38:40.585758-04:00 qeos-176 test test_message_46
> 2017-03-15T05:38:40.585880-04:00 qeos-176 test test_message_47
> 2017-03-15T05:38:40.585881-04:00 qeos-176 test test_message_48
> 2017-03-15T05:38:40.585963-04:00 qeos-176 test test_message_49
> 2017-03-15T05:38:40.585965-04:00 qeos-176 test test_message_50
> 2017-03-15T05:43:00.848824-04:00 qeos-176 test sage_after_37
> 2017-03-15T05:43:00.848832-04:00 qeos-176 test test_message_after_38
> 2017-03-15T05:43:00.848950-04:00 qeos-176 test test_message_after_39
> 2017-03-15T05:43:00.848952-04:00 qeos-176 test test_message_after_40
> 2017-03-15T05:43:00.849169-04:00 qeos-176 test test_message_after_41
> 2017-03-15T05:43:00.849171-04:00 qeos-176 test test_message_after_42
> 2017-03-15T05:43:00.855269-04:00 qeos-176 test test_message_after_43
> 2017-03-15T05:43:00.855272-04:00 qeos-176 test test_message_after_44
> 2017-03-15T05:43:00.855380-04:00 qeos-176 test test_message_after_45
> 2017-03-15T05:43:00.855382-04:00 qeos-176 test test_message_after_46
> 2017-03-15T05:43:00.855515-04:00 qeos-176 test test_message_after_47
> 2017-03-15T05:43:00.855517-04:00 qeos-176 test test_message_after_48
> 2017-03-15T05:43:00.855627-04:00 qeos-176 test test_message_after_49
> 2017-03-15T05:43:00.855633-04:00 qeos-176 test test_message_after_50
> ```
>
> Everything works fine, if I put "sleep" of 6-7 seconds between force "logrotate" and "for loop".
> Even with smaller sleep of few seconds, got the same results.
>
> My guess that there is some issue with the offset, as if I got smaller number of messages in the "Sanity" part. Let's say 20. I will get something like missing first 14 messages + part of the 15th.
>
> I was playing a little with it, and in my case this lost is happening only in the range of 1-69 sanity messages already in input file before logrotate.
> If I have more than 69 messages in "Sanity" part all messages after are logged without lost. Which was tricky as logs are usually bigger and seems that there is no problem then.
>
> Also I tried with polling and inotify mode., same issue.

The issue is that the only clue that rsyslog has in the copy-truncate scenario 
is if the file size shrinks between times that it looks at the file. If you 
truncate it and write to it immediatly, it isn't going to detect that the 
contents of the file before the point it's looking at have changed. In this 
case, you are writing longer messages to the file, so around message 37 the file 
becoems longer than it was before and rsyslog notices the new messages.

copy-truncate is guaranteed to loose messages, the only possible thing rsyslog 
could do to loose less is to check the file more frequently, but that wouldn't 
prevent message loss, only reduce it.

for your test, instead of writing 50 messaes, truncateing and repeating, write 
500 truncate and write 50, then rsyslog has a chance of detecting that something 
changed.

**Comment by 7echno** [comment]
Hi David,

Thank you for your answer, things are more clear now. And I manage to extend tests on this a little bit. So this are draft of my findings and observation.

There is this "interval" which is around 6 seconds, when pointer/offset is checked and set again.
So if you do logrotate, and inside this "interval" you write messages and manage to stay lower then previous set, with reopenOnTruncate="on" you will get all the messages.
If you hit the previous set, the size of the file, or grow more in size. 
You will loose the size of the previous offset.
Not going to happen often, but still it's easy to reproduce.

I get that HUP signal is just closing the open files, maybe it would be also a nice feature to force this set of offset/pointer to the end of file or something similar. This way this interval and copytrancate can be reduced significantly if it's set in logrotate config.

Thank you.

**Comment by rgerhards** [comment]
This is pretty complicated, and it doesn't work either ;-) The bottom line is that copytruncate should not be used, and there is no reliable way to support it. So let's not pretend we do.

**Comment by timp87** [comment]
@rgerhards  why did you close my ticket? My problem is unrelated to what @7echno decribed.

**Comment by rgerhards** [comment]
Sorry, I did not notice it was hijacked...

**Comment by davidelang** [comment]
@timp87 how are the logs getting rotated in your case?

**Comment by 7echno** [comment]
David, Rainer,

Thank you for you answers. Sorry for hijacking issue, was not my intention.

**Comment by timp87** [comment]
@davidelang everyhing I know it's a java process which uses Logback lib for rotation.

I took [this](https://pastebin.com/Axvf0EHS) simple program and caught inotify evens to see what's going on with log files on fs.
Here is the result, see evens for pas.main.log
https://drive.google.com/file/d/0By9yUmYFrRAKd19OeGN5NU94OGs/view?usp=sharing

Note, during this gathering process rsyslog-8.19 (w/o bug) was reading pas.main.log file as usual.

**Comment by timp87** [comment]
From user point of view it looks like app rotates its log, creates new one and archives the old one immediately

**Comment by davidelang** [comment]
Ok, that seems to show that it's renaming the file, so rsyslog should 
 be catching this

**Comment by kortemik** [comment]
Hello,

I think we are experiencing the same problem and therefore I played a little bit with the file monitor.

We found at least two cases where imfile at least on 8.27.0 works little bit off:

1) atomic renaming file, inside the same filesystem: rsyslog does not manage to catch it as a new file even if the new name matches the wildcard monitor path.

2) removing an monitored file which is still open: rsyslog removes the monitor for new events but still trakcs the handle.

3) I am not one hundred percent sure that the two above cases cause the one we are also seeing that some times new files go unnoticed.

Restart always helps. HUP? No effect!

**Comment by alorbach** [comment]
I am going to look into this very soon, if I can get a new testcase running that reproduces and tests the problem, we will have a good chance to find the rootcause of this problem.

**Comment by rgerhards** [comment]
@kortemik 
> Restart always helps. HUP? No effect!
Well, that's expected, because imfile is not affected by HUP. Did you find a doc that says it is? If so, I would like to correct it.

HUP closes files and reloads lookup tables. That's it.

**Comment by rgerhards** [comment]
I need to add that HUP has no effect because there should be no reason for it to have any effect.

**Comment by timp87** [comment]
@rgerhards, I've just updated to 8.28.0 and see no difference. The problem is still here.

**Comment by rgerhards** [comment]
@alorbach could you pls have a look

**Comment by alorbach** [comment]
I need more information. A new debug log would be helpful

**Comment by timp87** [comment]
@alorbach I need some time to catch it and gather more info

**Comment by timp87** [comment]
Same on 8.29. Still trying to figure out the way to reproduce it manually.

**Comment by alorbach** [comment]
Ok I am waiting for a reproduceable setup during meantime

**Comment by timp87** [comment]
Could anybody reopen this bug?
@alorbach, well, I can't form the actual steps to reproduce.
You can see my setup in the initial post here. IPs and DNS names is provided configs and rsyslog debug log can differ.

I managed to reproduce this again in our PRO:

Our applications uses logback for logging and rotation. Rsyslog uses imfile to read app logs.
And to fully understand what happened with pas.main.log and how it was rotated I installed inotify-tools and run inotifywait utility before, during and after rotation:
```
# inotifywait -m -r -o /root/inotifywait.`date "+%Y-%m-%d_%H:%M:%S"` --timefmt "%H:%M:%S" --format "%T | %w%f | %e"  /var/log/mydomain/pas
```
Here is the output https://drive.google.com/file/d/0By9yUmYFrRAKUGxsRld6Q29wMUU/view?usp=sharing

This is how rsyslog was started:
```
# env RSYSLOG_DEBUGLOG=/root/rsyslog.`date "+%Y-%m-%d_%H:%M:%S"`.debug RSYSLOG_DEBUG=Debug /sbin/rsyslogd -d -n -i /var/run/syslogd.pid > /dev/null 2>&1
```
Here is the output https://drive.google.com/file/d/0By9yUmYFrRAKalZoYlRSVGxjN28/view?usp=sharing

Please, note, I started and stopped rsyslog in debugging mode and inotifywait at the same time:
```
# date && pkill rsyslog && pkill notifywait
Tue Aug 29 12:18:23 UTC 2017
```

Rsyslog is tracking changes in pas.main.log. At about **12:15:14 UTC file was rotated.**
And this is how it looked like from lsof point of view after some time:
```
# lsof -p `pgrep rsyslog` | grep var/log                         Tue Aug 29 12:15:51 2017
COMMAND    PID USER   FD      TYPE              DEVICE  SIZE/OFF       NODE NAME
rsyslogd 18117 root   11w      REG              253,1    423936        157 /var/log/spool/rsyslog/queue-remote.00000002
rsyslogd 18117 root   12r      REG              253,1  10485995        153 /var/log/spool/rsyslog/queue-remote.00000001 (deleted)
rsyslogd 18117 root   13r      REG              253,1    423936        157 /var/log/spool/rsyslog/queue-remote.00000002
rsyslogd 18117 root   17w      REG              253,1   6661422        124 /var/log/syslog.log
rsyslogd 18117 root   18r      REG              253,1  15410631     819212 /var/log/mydomain/pas/pas.statistics-api.log
rsyslogd 18117 root   20w      REG              253,1  22020434        149 /var/log/messages
rsyslogd 18117 root   21w      REG              253,1  13091462         44 /var/log/kern.log
rsyslogd 18117 root   22r      REG              253,1   3152652     819203 /var/log/mydomain/pas/pas.main.log46920459489546.tmp (deleted)
rsyslogd 18117 root   24r      REG              253,1       317     819208 /var/log/mydomain/pas/pas.statistics-sms.log
rsyslogd 18117 root   28w      REG              253,1    126676        154 /var/log/secure
```
Rsyslog keeps open rotated and deleted pas.main.log46920459489546.tmp file with inode 819203.
Meanwhile new file was created and used by app with inode 819270:
```
# stat /var/log/mydomain/pas/pas.main.log
  File: ‘/var/log/mydomain/pas/pas.main.log’
  Size: 2958155         Blocks: 5784       IO Block: 4096   regular file
Device: fd01h/64769d    Inode: 819270      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/     pas)   Gid: ( 1000/     pas)
Access: 2017-08-29 12:15:14.664625925 +0000
Modify: 2017-08-29 12:15:56.973597729 +0000
Change: 2017-08-29 12:15:56.973597729 +0000
 Birth: -
```

I hope rsyslog.debug and inotifywatch output will help.

Look at inotifywatch output, It's visible that application moved pas.main.log to pas.main.log46920459489546.tmp, opened it, readed for gzipping, closed and then deleted it.
But there was no event from rsyslog that it closed this file too.

**Comment by josiahp** [comment]
I am noticing the same issue, also running 8.29.

I have a simple setup which reads files from a directory specified by wildcard and when new files are created rsyslog does not notice them until I restart or force-reload it.

This bug prevents me from using any recent of rsyslog in production, as we rely on newly generated files being delivered automatically.

**Comment by alorbach** [comment]
@josiahp Can you create n new issue for this case? 
This issue was about imfile not detecting a logrotation. And please post your configuration in the new issue as well, so I can try to setup and reproduce your problem.

**Comment by rgerhards** [comment]
I have locked the issue as it was fixed in v8.28 and is no longer being tracked. If with the current version an issue exists, please open a new issue. If you are 100% sure that what you see is exactly the same issue reported here, please let me know at rgerhards@adiscon.com

**Comment by rgerhards** [comment]
From @timp87 via mail:
```
Seems like my problem was fixed in 8.30.
At least I'm running this version since it was release and hasn't
faced with that bug.
Could you please leave a comment in
https://github.com/rsyslog/rsyslog/issues/1417 that initial reporter
is happy with rsyslog version 8.30?
```
Sorry for keeping this locked, but experience shows that such solved issues sometimes drag in many comments from folks who use an older version and that makes everyone's life miserable ;-)
